pkg/controller/warehouses/warehouses.go

```diff
@@ -284,6 +287,77 @@ func (r *reconciler) syncWarehouse(
 	// Automatically create a Freight from the latest discovered artifacts
 	// if the Warehouse is configured to do so.
 	if pol := warehouse.Spec.FreightCreationPolicy; pol == kargoapi.FreightCreationPolicyAutomatic || pol == "" {
+		criteriaSatisfied, err := freightCreationCriteriaSatisfied(ctx,
+			warehouse.Spec.FreightCreationCriteria,
+			status.DiscoveredArtifacts,
+		)
+		if err != nil {
+			logger.Error(err, "error evaluating freight creation criteria")
+			conditions.Set(
+				&status,
+				&metav1.Condition{
+					Type:   kargoapi.ConditionTypeHealthy,
+					Status: metav1.ConditionFalse,
+					Reason: "FreightCreationCriteriaError",
+					Message: fmt.Sprintf(
+						"failed to evaluate freight creation criteria: %s", err.Error(),
+					),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+			)
+			conditions.Set(
+				&status,
+				&metav1.Condition{
+					Type:   kargoapi.ConditionTypeReady,
+					Status: metav1.ConditionFalse,
+					Reason: "FreightCreationCriteriaError",
+					Message: fmt.Sprintf(
+						"failed to evaluate freight creation criteria: %s", err.Error(),
+					),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+			)
+			return status, fmt.Errorf("failed to evaluate freight creation criteria: %w", err)
+		}
+		if !criteriaSatisfied {
+			logger.Info("freight creation criteria not satisfied; skipping freight creation")
+			conditions.Set(
+				&status,
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeCriteriaSatisfied,
+					Status:             metav1.ConditionFalse,
+					Reason:             "FreightCreationCriteriaNotSatisfied",
+					Message:            "freight creation criteria not satisfied; skipping freight creation",
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeReconciling,
+					Status:             metav1.ConditionFalse,
+					Reason:             "FreightCreationCriteriaNotSatisfied",
+					Message:            "freight creation criteria not satisfied; skipping freight creation",
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
```

Comment:
I believe it's more idiomatic to delete the Reconciling condition if we're done.